File: backend/models/contract.py
from sqlalchemy import Column, Integer, String, DateTime, Text, Float, ForeignKey, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from database import Base


# 결제 수단 옵션 (프론트와 동일하게 유지)
PAYMENT_METHODS = [
    "bank_transfer",   # 계좌이체
    "card",            # 카드결제
    "government_aid",  # 정부지원금
]

# 계약 상태
CONTRACT_STATES = [
    "active",      # 진행 중
    "completed",   # 완료 (모두 입금 받음)
    "cancelled",   # 취소
]


class Contract(Base):
    __tablename__ = "contracts"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    customer_id = Column(
        Integer,
        ForeignKey("customers.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    estimate_id = Column(
        Integer,
        ForeignKey("saved_estimates.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    title = Column(String(300), nullable=False, default="")
    contract_amount = Column(Float, nullable=False, default=0)
    contract_date = Column(String(20), nullable=False, default="")
    state = Column(String(30), nullable=False, default="active")
    tax_invoice_issued = Column(Boolean, nullable=False, default=False)
    memo = Column(Text, nullable=False, default="")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    payments = relationship(
        "Payment",
        back_populates="contract",
        cascade="all, delete-orphan",
        order_by="Payment.paid_at",
    )


class Payment(Base):
    __tablename__ = "contract_payments"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    contract_id = Column(
        Integer,
        ForeignKey("contracts.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    amount = Column(Float, nullable=False, default=0)
    paid_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
    method = Column(String(30), nullable=False, default="bank_transfer")
    memo = Column(Text, nullable=False, default="")
    handler = Column(String(100), nullable=False, default="")  # 입금 받은 담당자
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    contract = relationship("Contract", back_populates="payments")

File: frontend/src/types/contract.ts
export type PaymentMethod = 'bank_transfer' | 'card' | 'government_aid';

export type ContractState = 'active' | 'completed' | 'cancelled';

export const PAYMENT_METHOD_LABELS: Record<PaymentMethod, string> = {
  bank_transfer: '계좌이체',
  card: '카드결제',
  government_aid: '정부지원금',
};

export const PAYMENT_METHOD_OPTIONS: PaymentMethod[] = [
  'bank_transfer', 'card', 'government_aid',
];

export const CONTRACT_STATE_LABELS: Record<ContractState, string> = {
  active: '진행 중',
  completed: '완료',
  cancelled: '취소',
};

export const CONTRACT_STATE_OPTIONS: ContractState[] = ['active', 'completed', 'cancelled'];

export const CONTRACT_STATE_COLORS: Record<ContractState, string> = {
  active: 'bg-blue-100 text-blue-700 border-blue-200',
  completed: 'bg-green-100 text-green-700 border-green-200',
  cancelled: 'bg-red-100 text-red-700 border-red-200',
};

export interface Payment {
  id: number;
  amount: number;
  paidAt: string;
  method: PaymentMethod;
  memo: string;
}

export interface PaymentInput {
  amount: number;
  paidAt?: string;
  method: PaymentMethod;
  memo: string;
}

export interface ContractInput {
  title: string;
  contractAmount: number;
  contractDate: string;
  estimateId?: number | null;
  state: ContractState;
  memo: string;
}

export interface ContractDetail {
  id: number;
  customerId: number;
  estimateId: number | null;
  title: string;
  contractAmount: number;
  contractDate: string;
  state: ContractState;
  memo: string;
  paidAmount: number;
  remainingAmount: number;
  createdAt: string;
  updatedAt: string;
  payments: Payment[];
}

export const emptyContract = (): ContractInput => ({
  title: '',
  contractAmount: 0,
  contractDate: new Date().toISOString().split('T')[0],
  state: 'active',
  memo: '',
});

export const emptyPayment = (): PaymentInput => ({
  amount: 0,
  paidAt: new Date().toISOString().slice(0, 16),
  method: 'bank_transfer',
  memo: '',
});
